File: GameServer/wwwroot/js/Physics/Vector2.js

function Vector2(x = 0, y = 0) {

    const that = this;

    that._x = x;
    that.GetX = function () {
        return that._x;
    };
    that.SetX = function(value) {

        that._x = value;
        that._length = undefined;
        that._lengthSquared = undefined;
    };

    that._y = y;
    that.GetY = function () {
        return that._y;
    };
    that.SetY = function (value) {
        
        that._y = value;
        that._length = undefined;
        that._lengthSquared = undefined;
    };

    that.Set = function (x, y) {

        that.SetX(x);
        that.SetY(y);
    };

    that.CopyFrom = function (vector) {

        that.SetX(vector.GetX());
        that.SetY(vector.GetY());
        that._length = vector._length;
        that._lengthSquared = vector._lengthSquared;
    };

    that.LengthSquared = function () {

        if (typeof that._lengthSquared === "undefined")
            that._lengthSquared = Math.pow(that.GetX(), 2) + Math.pow(that.GetY(), 2);

        return that._lengthSquared;
    };

    that.Length = function () {

        if (typeof that._length === "undefined")
            that._length = Math.sqrt(that.LengthSquared());

        return that._length;
    };

    that.Normalize = function () {

        if (that.IsDegenerated())
            throw "Degenerated vector given. Cannot normalize zero vector.";

        let vector = Vector2.Clone(that);
        vector = Vector2.Divide(vector, that.Length());
        return vector;
    };

    that.SafeNormalize = function () {

        try {
            that.Normalize();
        }
        catch (err) {
            return new Vector2();
        }
    };

    that.IsDegenerated = function () {

        return that.LengthSquared() === 0;
    };

    // Helpers
    that.Equals = function (otherVector) {

        if (!otherVector)
            return false;

        return that.GetX() === otherVector.GetX() && that.GetY() === otherVector.GetY();
    };

    that.ToString = function () {

        return `<${that.GetX()} ${that.GetY()}>`;
    };
};

Vector2.Clone = function(vector) {

    const copiedVector = new Vector2(vector.GetX(), vector.GetY());

    copiedVector._length = vector._length;
    copiedVector._lengthSquared = vector._lengthSquared;

    return copiedVector;
};

Vector2.Dot = function (vectorA, vectorB) {

    return (vectorA.GetX() * vectorB.GetX()) + (vectorA.GetY() * vectorB.GetY());
};

Vector2.Norm2 = function (vector) {

    const sumsqr = Math.pow(vector.GetX(), 2) + Math.pow(vector.GetY(), 2);
    return Math.sqrt(sumsqr);
};

Vector2.Similarity = function (vectorA, vectorB) {

    return Vector2.Dot(vectorA, vectorB) / Vector2.Norm2(vectorA) / Vector2.Norm2(vectorB);
};

Vector2.DistanceSquared = function (vectorA, vectorB) {

    return Math.pow(vectorA.GetX() - vectorB.GetX(), 2) + Math.pow(vectorA.GetY() - vectorB.GetY(), 2);
};

Vector2.Distance = function (vectorA, vectorB) {

    return Math.sqrt(Vector2.DistanceSquared(vectorA, vectorB));
};

Vector2.Normalize = function (vector) {

    return vector.Normalize();
};

Vector2.RadianToVector2 = function (radian) {

    return new Vector2(Math.cos(radian), Math.sin(radian));
};

Vector2.Vector2ToRadian = function (vector) {

    return Math.atan2(vector.GetY(), vector.GetX());
};

Vector2.AngleBetweenVectors = function (vectorA, vectorB) {

    if (vectorA.IsDegenerated() || vectorB.IsDegenerated())
        throw "One of given vectors is degenerated. Cannot calculate angle between them";

    const sin = vectorB.GetX() * vectorA.GetY() - vectorA.GetX() * vectorB.GetY();
    const cos = vectorB.GetX() * vectorA.GetX() + vectorB.GetY() * vectorA.GetY();

    return Math.atan2(sin, cos);
};

// Mathematical operations
Vector2.Add = function (vectorA, vectorB) {

    return new Vector2(vectorA.GetX() + vectorB.GetX(), vectorA.GetY() + vectorB.GetY());
};

Vector2.Subtract = function (vectorA, vectorB) {

    return new Vector2(vectorA.GetX() - vectorB.GetX(), vectorA.GetY() - vectorB.GetY());
};

Vector2.Multiply = function (vector, multiplier) {

    return new Vector2(vector.GetX() * multiplier, vector.GetY() * multiplier);
};

Vector2.Divide = function (vector, divider) {

    if (divider === 0)
        throw "Division by 0";

    return Vector2.Multiply(vector, 1.0 / divider);
};

Vector2.LEFT_VECTOR = () => new Vector2(-1, 0);
Vector2.RIGHT_VECTOR = () => new Vector2(1, 0);
Vector2.UP_VECTOR = () => new Vector2(0, -1);
Vector2.DOWN_VECTOR = () => new Vector2(0, 1);
Vector2.ZERO_VECTOR = () => new Vector2(0, 0);

// Export module
if (typeof module !== 'undefined' && module.hasOwnProperty('exports')) {
    module.exports = Vector2;
}

File: GameServer/Models/Vector2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GameServer.Models
{
    public class Vector2 : IEquatable<object>, ICloneable
    {
        /*
         * Vector logic
         */

        private double _x;
        public double X
        {
            get
            {
                return _x;
            }
            set
            {
                _lengthSquared = null;
                _length = null;

                _x = value;
            }
        }

        private double _y;
        public double Y
        {
            get
            {
                return _y;
            }
            set
            {
                _lengthSquared = null;
                _length = null;

                _y = value;
            }
        }

        public Vector2(double x = 0, double y = 0)
        {
            X = x;
            Y = y;
        }

        public Vector2(Vector2 vector)
            : this(vector.X, vector.Y)
        {
            _length = vector._length;
            _lengthSquared = vector._lengthSquared;
        }
        
        /*
         * Vector operations
         */
        private double? _lengthSquared = null;
        public double LengthSquared()
        {
            if (!_lengthSquared.HasValue)
                _lengthSquared = Math.Pow(X, 2) + Math.Pow(Y, 2);

            return _lengthSquared.Value;
        }

        private double? _length = null;
        public double Length()
        {
            if (!_length.HasValue)
                _length = Math.Sqrt(this.LengthSquared());

            return _length.Value;
        }

        public Vector2 Normalize()
        {
            if (IsDegenerated())
                throw new Exception("Degenerated vector given. Cannot normalzie zero vector.");

            var v = new Vector2(this);
            v /= Length();
            return v;
        }

        public Vector2 SafeNormalize()
        {
            try
            {
                return Normalize();
            }
            catch
            {
                return new Vector2();
            }
        }

        public bool IsDegenerated()
        {
            return LengthSquared() == 0;
        }

        /*
         * Static vector operations
         */
        public static double Dot(Vector2 a, Vector2 b)
        {
            return (a.X * b.X) + (a.Y * b.Y);
        }

        public static double DistanceSquared(Vector2 a, Vector2 b)
        {
            return Math.Pow(a.X - b.X, 2) + Math.Pow(a.Y - b.Y, 2);
        }

        public static double Distance(Vector2 a, Vector2 b)
        {
            return Math.Sqrt(DistanceSquared(a, b));
        }

        public static Vector2 Normalize(Vector2 v)
        {
            return v.Normalize();
        }

        public static Vector2 RadianToVector2(double radian)
        {
            return new Vector2(Math.Cos(radian), Math.Sin(radian));
        }

        public static double AngleBetweenVectors(Vector2 a, Vector2 b)
        {
            if (a.IsDegenerated() || b.IsDegenerated())
                throw new ArgumentException("One of given vectors is degenerated. Cannot calculate angle between them");

            var sin = b.X * a.Y - a.X * b.Y;
            var cos = b.X * a.X + b.Y * a.Y;

            return Math.Atan2(sin, cos);
        }

        /*
         * Mathematical operations
         */
        // Adding vectors
        public static Vector2 Add(Vector2 a, Vector2 b)
        {
            return new Vector2(a.X + b.X, a.Y + b.Y);
        }
        public static Vector2 operator +(Vector2 a, Vector2 b)
        {
            return Add(a, b);
        }

        // Subtracting vectors
        public static Vector2 Subtract(Vector2 a, Vector2 b)
        {
            return new Vector2(a.X - b.X, a.Y - b.Y);
        }
        public static Vector2 operator -(Vector2 a, Vector2 b)
        {
            return Subtract(a, b);
        }

        // Multiplying vectors
        public static Vector2 Multiply(Vector2 a, double b)
        {
            return new Vector2(a.X * b, a.Y * b);
        }
        public static Vector2 operator *(Vector2 a, double b)
        {
            return Multiply(a, b);
        }

        // Dividing vectors
        public static Vector2 Divide(Vector2 a, double b)
        {
            if (b == 0)
                throw new ArgumentException("Division by 0");

            return Multiply(a, 1.0 / b);
        }
        public static Vector2 operator /(Vector2 a, double b)
        {
            return Divide(a, b);
        }

        // Comparing vectors
        public static bool operator ==(Vector2 a, Vector2 b)
        {
            return a.Equals(b);
        }
        public static bool operator !=(Vector2 a, Vector2 b)
        {
            return !a.Equals(b);
        }

        public override bool Equals(object obj)
        {
            if (obj is null)
                return false;

            var v = obj as Vector2;
            if (v == null)
                return false;

            return (X == v.X && Y == v.Y);
        }

        public override string ToString()
        {
            return $"<{X} {Y}>";
        }

        public override int GetHashCode()
        {
            unchecked // Overflow is fine, just wrap
            {
                int hash = 17;
                hash = hash * 23 + X.GetHashCode();
                hash = hash * 23 + Y.GetHashCode();
                return hash;
            }
        }
        
        public object Clone()
        {
            return new Vector2(this);
        }

        /*
         * Predefined direction vectors
         */
        public static readonly Vector2 UP_VECTOR = new Vector2(0, -1);
        public static readonly Vector2 DOWN_VECTOR = new Vector2(0, 1);
        public static readonly Vector2 LEFT_VECTOR = new Vector2(-1, 0);
        public static readonly Vector2 RIGHT_VECTOR = new Vector2(1, 0);
        public static readonly Vector2 ZERO_VECTOR = new Vector2(0, 0);
    }
}
